### PR TITLE
added spliced mRNA output if sequence is of type mRNA

### PR DESCRIPTION
**Pull request by mdondrup**
I have added a section to the sequence template that outputs aggregated (spliced)
exon sequence for sequences of type mRNA.

          

Issue #

## Type(s) of Change(s)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API-specific change (fix or addition to an API function)
- [ ] Updates documentation (inline or markdown files)

## Description

There was no straight forward way to access the spliced mRNA sequence for a transcript. 
Users wish for a way to retrieve the transcript sequence without introns via copy-paste.
The "from alignment" section shows exon locations highlighted, but to get the actual sequence 
was a bit cumbersome.

## Testing?

Tested on our test site, then moved into the production site.

PHP 7.1

It only affects sequences of type mRNA, should not affect other code.

## Additional Notes (if any):

- [?] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by spficklin** [comment]
Hi @mdondrup thanks for sharing!   I'm a bit occupied with pull requests for our imminent Tripal v3 release.  So, I can get to this once that settles a bit.

**Comment by bradfordcondon** [comment]
hi @mdondrup The testing section is for describing how others could check out the feature/fix.   Do you have a GFF file (ideally just a snippet of it) that I could load to create the sort of feature structure this addresses? And/or more information on how to load data to create this issue?  I'm guessing i would load a GFF with exons/mrna/genes defined and then the FASTA sequences for.... the genes?  

I've generally loaded the CDS FASTA sequences associated with the mRNA so I haven't encountered this problem.  

ps thanks for the contribution!

**Comment by mdondrup** [comment]
Hi,

here is a small test-file. I would like to note that in general CDS != mRNA sequence, and therefore both are needed. We have imported GFF files using the Chado-perl (gmod_bulk_load_gff3) importer. There could be a minor difference in handling CDS between the Chado importer and the Tripal gff importer.  


[tking-test.gff3.txt](https://github.com/tripal/tripal/files/2427452/tking-test.gff3.txt)

**Comment by bradfordcondon** [comment]
thanks @mdondrup ! For others looking to test, you can manually create two landmark features (necessary prior to importing): `scaffold2972_size10933` and `scaffold342_size201858`

**Comment by bradfordcondon** [comment]
hi @mdondrup first of all im sorry this is taking so long to get reviewed. 

I get this error trying to load your GFF:

```

There are 1 jobs queued.
Calling: tripal_feature_load_gff3(/var/www/html/sites/default/files/tking-test.gff3.txt, 1, 1, 0, 1, 0, 0, 1, , , 0, , , , 0, , , 37)

NOTE: Loading of this GFF file is performed using a database transaction.
If the load fails or is terminated prematurely then the entire set of
insertions/updates is rolled back and will not be found in the database

Opening /var/www/html/sites/default/files/tking-test.gff3.txt
WD tripal_feature: Attribute is not correctly formatted on line 7: C:integral to membrane                                                          [error]
[site http://default] [TRIPAL ERROR] [TRIPAL_FEATURE] Attribute is not correctly formatted on line 7: C:integral to membrane
```

looks like the colons in the Note="" are messing it up.  I guess ill just remove that text since i dont think its necessary for testing this.

**Comment by mdondrup** [comment]
Hi,

I am sorry. I remember that this file required several tweaks to import correctly and finally I used
the chado perl importer. Removing the Note attribute should be fine.

Michael

File: tripal_feature/theme/templates/tripal_feature_sequence.tpl.php [hunk 1]
```diff
@@ -52,7 +52,32 @@
...
   }
-  
+ 
+   // provide spliced mRNA sequences
+   // if this is an mRNA
+  if ($feature->type_id->name == 'mRNA') {
+
+
+      $seqs = tripal_get_feature_sequences(array('feature_id' => $feature->feature_id), array('derive_from_parent' => 1, 'aggregate' => 1,
+       sub_feature_types => array('exon') ,'is_html' => 1, 'width' => $num_bases));
```
Comment (bradfordcondon) [inline tripal_feature/theme/templates/tripal_feature_sequence.tpl.php]: should this be 'sub_feature_types' ? note quotes.  I get an undefined constant warning.

File: tripal_feature/theme/templates/tripal_feature_sequence.tpl.php [hunk 2]
```diff
@@ -52,7 +52,32 @@
...
+      //echo( ">". $seqs[0]['names']['header']. " \n".($seqs[0]['residues'])."\n" );
+      $list_items[] = '<a href="#mRNA">'. $feature->type_id->name . ' - spliced transcript sequence  ' . "</a>";
+      $sequences_html .= '<a name="mRNA"></a>';
+      $sequences_html .= '<div id="mRNA" class="tripal_feature-sequence-item">';
+      $sequences_html .= '<p><b>Spliced messenger RNA</b></p>';
+      $sequences_html .= '<pre class="tripal_feature-sequence">';
+      $sequences_html .= ">". $seqs[0]['defline']. " \n".$seqs[0]['residues'];
```
Comment (bradfordcondon) [inline tripal_feature/theme/templates/tripal_feature_sequence.tpl.php]: What if there are no seqs?    Can we hide this whole display in that case?

**Review by bradfordcondon — CHANGES_REQUESTED** [review]
Thank you for this submission!

I get these warnings:


```
Notice: Use of undefined constant sub_feature_types - assumed 'sub_feature_types' in include() (line 62 of /var/www/html/sites/all/modules/custom/tripal/tripal_feature/theme/templates/tripal_feature_sequence.tpl.php).
Notice: Undefined offset: 0 in include() (line 69 of /var/www/html/sites/all/modules/custom/tripal/tripal_feature/theme/templates/tripal_feature_sequence.tpl.php).
Notice: Undefined offset: 0 in include() (line 69 of /var/www/html/sites/all/modules/custom/tripal/tripal_feature/theme/templates/tripal_feature_sequence.tpl.php).
```

* I didn't set up my features like you did, and thats why i get an empty field with some warnings.  What should i have the residues associated with to get this to display?  The contig  feature?

 *  I think that this field should only display if its able to display the spliced sequence.  This should be Easy to do by checking if `$seqs = tripal_get_feature_sequences()` returns anything or not, and only appending to `$sequences_html` if it does.

screen shot 2018-10-05 at 12 35 58 pm


PS thanks for the submssion and for your patience....

**Comment by mdondrup** [comment]
> On Oct 5, 2018, at 6:34 PM, Bradford Condon <notifications@github.com> wrote:
> 
> @bradfordcondon commented on this pull request.
> 
> In tripal_feature/theme/templates/tripal_feature_sequence.tpl.php:
> 
> > @@ -52,7 +52,32 @@
>      $sequences_html .= '</div>';
>      
>    }
> -  
> + 
> +   // provide spliced mRNA sequences
> +   // if this is an mRNA
> +  if ($feature->type_id->name == 'mRNA') {
> +
> +
> +      $seqs = tripal_get_feature_sequences(array('feature_id' => $feature->feature_id), array('derive_from_parent' => 1, 'aggregate' => 1,
> +       sub_feature_types => array('exon') ,'is_html' => 1, 'width' => $num_bases));
> 
> should this be ‘sub_feature_types' ?

Yes, I guess in PHP array index should be quoted.

**Comment by mdondrup** [comment]
> On Oct 5, 2018, at 6:35 PM, Bradford Condon <notifications@github.com> wrote:
> 
> @bradfordcondon commented on this pull request.
> 
> In tripal_feature/theme/templates/tripal_feature_sequence.tpl.php:
> 
> > -  
> + 
> +   // provide spliced mRNA sequences
> +   // if this is an mRNA
> +  if ($feature->type_id->name == 'mRNA') {
> +
> +
> +      $seqs = tripal_get_feature_sequences(array('feature_id' => $feature->feature_id), array('derive_from_parent' => 1, 'aggregate' => 1,
> +       sub_feature_types => array('exon') ,'is_html' => 1, 'width' => $num_bases));
> +      //echo( ">". $seqs[0]['names']['header']. " \n".($seqs[0]['residues'])."\n" );
> +      $list_items[] = '<a href="#mRNA">'. $feature->type_id->name . ' - spliced transcript sequence  ' . "</a>";
> +      $sequences_html .= '<a name="mRNA"></a>';
> +      $sequences_html .= '<div id="mRNA" class="tripal_feature-sequence-item">';
> +      $sequences_html .= '<p><b>Spliced messenger RNA</b></p>';
> +      $sequences_html .= '<pre class="tripal_feature-sequence">';
> +      $sequences_html .= ">". $seqs[0]['defline']. " \n".$seqs[0]['residues'];
> 
> What if there are no seqs? Can we hide this whole display in that case?
> 

Yes, should I change that myself?

**Comment by bradfordcondon** [comment]
hi @mdondrup yes please! if you could push the change to your mdondrup fork and 7.x-2.1-licebase branch ( the source of this PR)  then it will auto-update this PR and I can retest.  PLus that way you keep credit for the work :)  

Also in case you missed it, if you could help me out and tell me which featue should have the residues so that this field displays properly i'd appreciate it (mrna? gene? contig? the CDS?) so that I can confirm the functionality on my test site :)

**Comment by bradfordcondon** [comment]
hi @mdondrup just following up on this, wondering if you needed any assistance.

Thanks!

**Comment by spficklin** [comment]
Hi @mdondrup this PR is getting really old.  In order to keep our PR list clean we would like to close this out.  If you can respond to @bradfordcondon in the next two weeks we can keep this open. Otherwise, we'll close it out and request you resubmit again.  We really appreciate the fix and hope you can get back on this one.

**Comment by mdondrup** [comment]
Hi. Sorry for the delayed response, was pretty busy with different stuff. I think I can fix it before new year. 

Cheers 
Michael

**Comment by spficklin** [comment]
Thanks @mdondrup.

File: tripal_feature/theme/templates/tripal_feature_sequence.tpl.php (hunk 1 above)
Comment (mdondrup) [reply tripal_feature/theme/templates/tripal_feature_sequence.tpl.php]: Fixed in commit 338342a1377463fbfa7d16baf793ed6fd8d0e433

File: tripal_feature/theme/templates/tripal_feature_sequence.tpl.php (hunk 2 above)
Comment (mdondrup) [reply tripal_feature/theme/templates/tripal_feature_sequence.tpl.php]: Fixed in commit 338342a1377463fbfa7d16baf793ed6fd8d0e433

**Comment by mdondrup** [comment]
Hi @bradfordcondon please have a look at commit 338342a1377463fbfa7d16baf793ed6fd8d0e433
I have added quotes and check for empty sequence results, at which the block is skipped.

**Comment by bradfordcondon** [comment]
OK I can confirm this is no longer throwing errors on a blank feature with no children, or with children without sequences.

However I cant reproduce the functional addition. Because I've never used the Chado perl loader, I just created an mRNA feature, and two exon features, with a relationship of subject = exon, object = mRNA, type = derives_from, and gave hte exon sequences....


ah, i need `derive_from_parent`.  should it also be `derives_from`?  I dont know much about Tripal 2: which relationship does the Tripal loader use?  I also dont have `derive_from_parent` in my terms table, is it a synonym?


edit: oops, derive_from_parent is a tripal key not a sequence ontology term.  I'll take another look....

**Comment by bradfordcondon** [comment]
@mdondrup In order for the spliced mRNA to show up on my site, I need to use `derives_from` instead of `derive_from_parent`. Searching for this term, I can't find it in the current sequence ontology.  Is it a custom or depracated term?


We can satisfy me by checking if the term is `derive_from_parent` OR `derives_from` since we do check that the type = exon.

Still I'd like to know where `derive_from_parent` comes from.

**Review by bradfordcondon — CHANGES_REQUESTED** [review]
hold on, i realize my mistake, `derive_from_parent` isnt an ontology term but a key passed into tripal, let me have another look.

Looks like `derive_from_parent` comes from featureloc, so i need to insert that info.  dang i dont think i can shortcut, i have to use the loader.

**Comment by bradfordcondon** [comment]
sorry about the mixup.  looks good, thank you.

**Pull request merged**